From f7d234914d4d8cfa46f328f053c7716bf91d5588 Mon Sep 17 00:00:00 2001
From: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Date: Mon, 22 Oct 2018 20:30:51 -0700
Subject: [PATCH 0485/1852] xilinx: v4l: scd: Implement generic event
 notification

commit 6878e089e5268351e9d867e07fc57cd217637b7c from
https://github.com/Xilinx/linux-xlnx.git

This patch implements generic SCD event notification function.

Signed-off-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../xilinx/xilinx-scenechange-channel.c       | 46 ++++++++++++-------
 .../platform/xilinx/xilinx-scenechange.h      |  1 -
 2 files changed, 29 insertions(+), 18 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
index 6a86213d3a14..aa1b98d38a90 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
@@ -41,6 +41,7 @@
 #define XSCD_STRIDE_OFFSET		0x20
 #define XSCD_VID_FMT_OFFSET		0x28
 #define XSCD_SUBSAMPLE_OFFSET		0x30
+#define XSCD_SAD_OFFSET			0x38
 
 /* Hardware video formats for memory based IP */
 #define XSCD_COLOR_FMT_Y8		24
@@ -55,6 +56,9 @@
 #define XSCD_V_SUBSAMPLING		16
 #define XSCD_BYTE_ALIGN			16
 
+#define XSCD_SCENE_CHANGE		1
+#define XSCD_NO_SCENE_CHANGE		0
+
 /* -----------------------------------------------------------------------------
  * V4L2 Subdevice Pad Operations
  */
@@ -311,29 +315,37 @@ static const struct media_entity_operations xscd_media_ops = {
 	.link_validate = v4l2_subdev_link_validate,
 };
 
+static void xscd_event_notify(struct xscd_chan *chan)
+{
+	u32 *eventdata;
+	u32 sad;
+
+	sad = xscd_read(chan->iomem, XSCD_SAD_OFFSET +
+			(chan->id * XILINX_XSCD_CHAN_OFFSET));
+	sad = (sad * 16) / (chan->format.width * chan->format.height);
+	eventdata = (u32 *)&chan->event.u.data;
+
+	if (sad >= 1)
+		eventdata[0] = XSCD_SCENE_CHANGE;
+	else
+		eventdata[0] = XSCD_NO_SCENE_CHANGE;
+
+	chan->event.type = V4L2_EVENT_XLNXSCD;
+	v4l2_subdev_notify_event(&chan->subdev, &chan->event);
+}
+
 static irqreturn_t xscd_chan_irq_handler(int irq, void *data)
 {
 	struct xscd_chan *chan = (struct xscd_chan *)data;
-	u32 sad;
-	u32 *eventdata;
+	struct xscd_shared_data *shared_data;
 
+	shared_data = (struct xscd_shared_data *)chan->dev->parent->driver_data;
 	spin_lock(&chan->dmachan.lock);
-	if (chan->dmachan.valid_interrupt) {
+
+	if ((shared_data->memory_based && chan->dmachan.valid_interrupt) ||
+	    !shared_data->memory_based) {
 		spin_unlock(&chan->dmachan.lock);
-		sad = xscd_read(chan->iomem, XILINX_XSCD_SAD_OFFSET +
-				(chan->id * XILINX_XSCD_CHAN_OFFSET));
-		sad = (sad * 16) / (chan->format.width * chan->format.height);
-		memset(&chan->event, 0, sizeof(chan->event));
-		eventdata = (u32 *)&chan->event.u.data;
-
-		if (sad >= 1)
-			eventdata[0] = 1;
-		else
-			eventdata[0] = 0;
-
-		chan->event.type = V4L2_EVENT_XLNXSCD;
-		v4l2_subdev_notify(&chan->subdev, V4L2_DEVICE_NOTIFY_EVENT,
-				   &chan->event);
+		xscd_event_notify(chan);
 		return IRQ_HANDLED;
 	}
 
diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.h b/drivers/media/platform/xilinx/xilinx-scenechange.h
index 618b9d42ffa3..72c6fc410ed3 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.h
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.h
@@ -31,7 +31,6 @@
 
 /* Register/Descriptor Offsets */
 #define XILINX_XSCD_ISR_OFFSET		0x0c
-#define XILINX_XSCD_SAD_OFFSET		0x38
 #define XILINX_XSCD_CHAN_OFFSET		0x100
 
 /* Interrupt Status and Control */
-- 
2.31.1

